lib/membrane_opus/parser/frame_lengths.ex

```diff
@@ -0,0 +1,191 @@
+      end)
+
+    frame_lengths =
+      if self_delimiting? do
+        lengths
+      else
+        last_frame_length =
+          byte_size(data_without_headers) - byte_offset - Enum.sum(lengths) - padding_length
+
+        lengths ++ [last_frame_length]
+      end
+
+    # adding 2 for TOC and code three header
+    {frame_lengths, 2 + byte_offset}
+  end
+
+  @spec code_three_cbr_lengths(binary, 0..1, non_neg_integer, boolean) ::
+          {[non_neg_integer], pos_integer}
+  defp code_three_cbr_lengths(data_without_headers, padding_flag, frame_count, self_delimiting?) do
+    {padding_length, padding_encoding_length} =
+      calculate_padding_info(padding_flag, data_without_headers)
+
+    {frame_duration, header_length} =
+      if self_delimiting? do
+        {frame_duration, bytes_used} =
+          calculate_frame_length(data_without_headers, padding_encoding_length)
+
+        # adding 2 for TOC and code three header
+        {frame_duration, 2 + padding_encoding_length + bytes_used}
+      else
+        frame_duration =
+          div(
+            byte_size(data_without_headers) - padding_encoding_length - padding_length,
+            frame_count
+          )
+
+        # adding 2 for TOC and code three header
+        {frame_duration, 2 + padding_encoding_length}
+      end
+
+    frame_lengths =
+      0..(frame_count - 1)
+      |> Enum.map(fn _i -> frame_duration end)
+
+    {frame_lengths, header_length}
+  end
+
+  # calculates frame length of the frame starting at byte_offset, specifically
+  # for code 2 or 3 packets, and the number of bytes used to encode the frame
+  # length
+  @spec calculate_frame_length(binary, non_neg_integer) :: {non_neg_integer, 1..2}
+  defp calculate_frame_length(data, byte_offset) do
+    <<_head::binary-size(byte_offset), length::size(8), _rest::binary>> = data
+
+    if length < 252 do
+      {length, 1}
+    else
+      <<_head::binary-size(byte_offset), _length::size(8), overflow_length::size(8),
+        _rest::binary>> = data
+
+      # https://tools.ietf.org/html/rfc6716#section-3.1
+      {overflow_length * 4 + length, 2}
+    end
+  end
+
+  # calculates total packet padding length, specifically for code 3 packets,
+  # and the number of bytes used to encode the padding length
+  @spec calculate_padding_info(0, any) :: {0, 0}
+  defp calculate_padding_info(padding_flag, _data) when padding_flag == 0, do: {0, 0}
+
+  @spec calculate_padding_info(1, binary) :: {non_neg_integer, non_neg_integer}
+  defp calculate_padding_info(padding_flag, data) when padding_flag == 1 do
+    do_calculate_padding_info(data)
+  end
+
+  @spec do_calculate_padding_info(binary, non_neg_integer, non_neg_integer) ::
+          {non_neg_integer, non_neg_integer}
+  defp do_calculate_padding_info(data, current_padding \\ 0, byte_offset \\ 0) do
+    <<padding::size(8), rest::binary>> = data
+
+    if padding == 255 do
+      do_calculate_padding_info(rest, current_padding + 254, byte_offset + 1)
```

Comment:
maybe a dumb question...but will this still trigger tail call optimization? i think it should because it's the last evaluated line, even though it's not the last written line. just wanna check.

Reply:
If it's the last call, the next instruction is `return`, so it's easy for the VM to find out that there's no need to leave the stack frame, thus the tail optimization will work. It doesn't matter if it's the last written line or not.